firestore/firestore_snippets/solution_counters.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strconv"
+	"time"
+
+	"cloud.google.com/go/firestore"
+	"google.golang.org/api/iterator"
+)
+
+// [START fs_counter_classes]
+// counters/${ID}
+type Counter struct {
+	numShards int
+}
+
+// counters/${ID}/shards/${NUM}
+type Shard map[string]int
+
+// [END fs_counter_classes]
+
+// [START fs_create_counter]
+func (d *Counter) сreateCounter(ctx context.Context, docRef *firestore.DocumentRef) []error {
```

Comment:
`createCounter` sounds like it will initialize a value of type `Counter`. Maybe this should return a `Counter` instead of receiving one? Also, if I'm trying to use this API, do I need to pass around Firestore references? Can we encapsulate all of the Firestore logic within the functions, rather than making the user handle and pass around the references?

Reply:
Divergence, other samples uses doc reference arg

Reply:
Got it. Could it return a `Counter`? OK if you think it should, but it's surprising to me the function is "Create" but doesn't create the `Counter`.

Reply:
I was concerned only about divergence. If it is not so strict, let's just rename method to `initCounter`. Actually, I'm on your side with this - create should return something